src/renderers/testing/ReactShallowRenderer.js

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright 2013-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule ReactShallowRenderer
+ * @preventMunge
+ */
+
+'use strict';
+
+const React = require('react');
+const ReactTestRenderer = require('react-test-renderer');
+
+const emptyFunction = require('fbjs/lib/emptyFunction');
+const invariant = require('fbjs/lib/invariant');
+
+const ShallowNodeMockComponent = ({children}) =>
+  React.Children.toArray(children);
+
+function createShallowNodeMock() {
+  let isFirst = true;
+  return function createNodeMock(element) {
+    if (isFirst) {
+      isFirst = false;
+      return element.type;
+    }
+    return ShallowNodeMockComponent;
+  };
+}
+
+function wrapElementWithContextProvider(element, context) {
+  const childContextTypes = Object.keys(context).reduce(
+    (reduced, key) => {
+      reduced[key] = emptyFunction;
+      return reduced;
+    },
+    {},
+  );
+
+  class ShallowRendererWrapper extends React.Component {
+    static childContextTypes = childContextTypes;
+    getChildContext() {
+      return context;
+    }
+    render() {
+      return this.props.children;
+    }
+  }
+
+  return React.createElement(ShallowRendererWrapper, null, element);
+}
+
+class ReactShallowRenderer {
+  getMountedInstance() {
+    return this._renderer ? this._renderer.getInstance() : null;
+  }
+
+  getRenderOutput() {
+    if (this._renderer) {
+      const tree = this._renderer.toTree();
+      if (tree && tree.rendered) {
+        // If we created a context-wrapper then skip over it.
+        const element = tree.type.childContextTypes
+          ? tree.rendered.rendered
+          : tree.rendered;
+
+        // Convert the rendered output to a ReactElement.
+        // This supports .toEqual() comparison for test elements.
+        return React.createElement(
```

Comment:
This would only re-trigger if the `ReactElement` returned from the top-level render was a class component with `propTypes`. In that case, it still wouldn't show a duplicate message because `checkPropTypes` tracks `loggedTypeFailures` and avoids re-logging duplicates.

Do you think this is reasonable? Should I look into ways of bypassing `checkPropTypes` for this case? (Or something else?)